File: Matlab Libraries/find_chirps.m
function find_chirps(data_path,mic_num,bat_type)
%   
%   This is a function that will systematically create a spectrogram of the
%   chirps that are detected from a specific dataset that is passed to the
%   function. This uses a filtered dataset for the time_domain_finder and
%   an unfiltered dataset for the frequency domain finder. Both sections do
%   not output filtered spectrograms, only unfiltered spectrograms are
%   used. 
% 
%   Sam Kramer
%   July 28th, 2023
%

%%%%%%%%%%%%%%%%%%%%%%%%%% Selecting Data Type %%%%%%%%%%%%%%%%%%%%%%%%%%%%

% --Workspace messages
    clc; warning off;

% --Load in data
    [data,~,fs] = load_data(data_path);
    disp("Loaded data successfully.")

% --Select Microphone out of data
    mic_num = uint8(mic_num);
    data = data(:,mic_num);

% --Select Bat type
    bat_type = string(bat_type);
    bat_type = char(upper(bat_type));
    try
        if bat_type(1) == 'H'
            % --Load in hippo call
                hippo_call = load("Hippo_example_chirp.mat");
                CFFM_chirp = hippo_call.model_chirp;
                disp("Selected Hipposideros")

            % --create linear chirp pattern
                FM_chirp = create_chirp("Linear",1,1,120000,100000,500000,"false");
                weight = 1.3;

        elseif bat_type(1) == 'R'
            % --Load in rhino calls
                rhino_call = load("Rhino_example_chirp.mat");
                CFFM_chirp = rhino_call.model_chirp;
                FM_chirp = create_chirp("Linear",1,1,120000,100000,500000,"false");
                weight = 0;
        end
    catch
        % --Error handler
            disp("No example chirp selected")
            disp("Checking only on a linear chirp pattern")

        % --create linear chirp pattern
            FM_chirp = create_chirp("Linear",1,1,120000,100000,500000,"false");
    end

% --Feedback message
    fprintf("Finding Chirps...\n")
    fprintf("This may take a moment.\n")

% --Filter data for correlations
    try
        filtered_data = filter_data(data,fs,100000,0.6,"false");
    catch
        filtered_data = filter_data(data,fs,[],[],"false");
            disp("Error with filter")
            disp("Ensure that Nyquist is above 200kHz")
    end

%%%%%%%%%%%%%%%%%%%%%%%%%%% Time Domain Section %%%%%%%%%%%%%%%%%%%%%%%%%%%

% Ghost Code:
    % --Compare the two time domain signals {Call time_domain_finder()}
        time_indeces = time_domain_finder(filtered_data,fs,FM_chirp,CFFM_chirp,weight);
        
    % --Outline for time domain section
        fprintf("Time Domain Finder Done.\n")
        
%%%%%%%%%%%%%%%%%%%%%%%%%%%% Freq Domain Section %%%%%%%%%%%%%%%%%%%%%%%%%%

% Ghost Code:
    % --Compare two frequency domain signals {Call freq_domain_fnder()}
        
    
    % --Outline for time domain section
        fprintf("Frequency Domain Finder Done.\n")

%%%%%%%%%%%%%%%%%%%%%%%%%%% Output Spectrograms %%%%%%%%%%%%%%%%%%%%%%%%%%%

% Ghost Code:
    % --Compare the two finder indeces


     % --Make new directory
        if ~exist("C:\Users\FIT UBD\Desktop\Data Analysis Spectrogram Files\","dir")
            cd("C:\Users\FIT UBD\Desktop")
            mkdir("Data Analysis Spectrogram Files")
            cd("C:\Users\FIT UBD\Desktop\Data Analysis Spectrogram Files")
        else
            cd("C:\Users\FIT UBD\Desktop\Data Analysis Spectrogram Files\")
        end

    % --Generate Spectrograms
        for i = 1:length(time_indeces)
            
            % --Call on spectrogram_function
                spect = figure('Visible','off');
                spectrogram_function(filtered_data,fs,time_indeces(i) - 0.021, time_indeces(i) + 0.021);

            % --Create title
                title = "Spectrogram" + string(i) + ".jpg";

            % --Save spectrogram of data
                saveas(spect,title)
                
        end

    % --Close out of the directory
        cd("C:\Users\FIT UBD\Desktop")
        fclose("all");
        close all;          % Close all the open figures
end

File: Julia Libraries/find_chirps.jl
#=  find_chirps()
    This is the main find chirps function that is going to be run to structure the algorithm within
    julia. This is going to be largely based on the matlab function that I have made but in a
    faster more flexible programming language. 

    Sam Kramer
    August 30th, 2023

=#

# --Include statements
    include("library_module.jl")

# --Using Statements
    using GLMakie
    using FFTW
    using SignalAnalysis
    using MAT

# --module inclusion
    using Main.library

############################################ Parameters ###########################################
    file_path = "C:\\Users\\Sam Kramer\\Desktop\\Engineering\\IRES 2023\\IRES 2023 -- Acoustics Array\\Test Data\\Testing 20JUL23\\Bat1_Trial2_20JUL2023.mat";
    mic_num = 1;
    bat_type = "Hipposiderus";

##################################### Function Definition #########################################

function find_chirps(file_path, mic_num, bat_type)

    # --Load in dataset and select mic dataset
        (mic_data, time, fs) = load_data(file_path);
        mic_data = vec(mic_data[:, mic_num]);

    # --Select Bat Call Comparison
        bat_type = String(bat_type);
        bat_type = uppercase(bat_type);

        try
            if bat_type[1] == 'H'
                # --Load in Hipposiderus call
                    Hippo_data = "C:\\Users\\Sam Kramer\\Desktop\\Engineering\\Chirp Detection Algorithm\\Matlab Libraries\\Hippo_example_chirp.mat";
                    CFFM_chirp = matread(Hippo_data);
                    CFFM_chirp = vec(CFFM_chirp["model_chirp"]);

            elseif bat_type[1] == 'R'
                # --Load in Rhinolophus call
                    Rhino_data = "C:\\Users\\Sam Kramer\\Desktop\\Engineering\\Chirp Detection Algorithm\\Matlab Libraries\\Rhino_example_chirp.mat";
                    
                    CFFM_chirp = matread(Rhino_data);
                    CFFM_chirp = vec(CFFM_chirp["model_chirp"]);

            end
        catch
            println("Error Loading Template Data: Testing Only Against Linear Chirp")
        end

    # --Create the FM_chirp
        FM_chirp = create_chirp(120000, 100000, 0.05, fs, "false");

    # --Feedback message
        println("Chirp Finder Beginning")
        println("This May Take a Moment...")

    # --Filter data out {Calls filter_data()}
        filtered_data = filter_data(data, fs);

    ####################################### Time Domain Section #######################################

    # --Create initial guesses in time domain {Calls time_domain_finder()}
        time_indeces = time_domain_finder(filtered_data, CFFM_chirp, FM_chirp);

    # --Outline confirmation message
        println("Time Domain Finder Done")

    ####################################### Freq Domain Section #######################################

    # --Create guesses from frequency domain methods {Calls freq_domain_finder()}


    # --Outline confirmation message
        println("Frequency Domain Finder Done")

    ####################################### Create Spectrograms #######################################

    # --Compare datasets


    # --Output and Save spectrograms


    # --Outline confirmation message
        println("Finder Done")

end     # End function